core/aws_ddk_core/pipelines/state_machine_stage.py

```diff
@@ -0,0 +1,157 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+from typing import List, Optional
+
+from aws_cdk.aws_events import EventPattern, IRuleTarget
+from aws_cdk.aws_events_targets import SfnStateMachine
+from aws_cdk.aws_stepfunctions import IChainable, StateMachine, StateMachineType
+from aws_cdk.aws_stepfunctions_tasks import EventBridgePutEvents, EventBridgePutEventsEntry
+from aws_ddk_core.pipelines.stage import DataStage
+from aws_ddk_core.resources import StepFunctionsFactory
+from constructs import Construct
+
+
+class StateMachineStage(DataStage):
+    """
+    DataStage, but contains helper methods for creating state machines and events.
+    """
+
+    def __init__(
+        self,
+        scope: Construct,
+        id: str,
+        name: Optional[str] = None,
+        description: Optional[str] = None,
+    ) -> None:
+        """
+        Create a stage.
+
+        Parameters
+        ----------
+        scope : Construct
+            Scope within which this construct is defined
+        id : str
+            Identifier of the stage
+        name : Optional[str]
+            Name of the stage
+        description :  Optional[str]
+            Description of the stage
+        """
+        super().__init__(scope, id)
+
+    def create_state_machine(
+        self,
+        id: str,
+        environment_id: str,
+        definition: IChainable,
+        state_machine_type: StateMachineType = StateMachineType.EXPRESS,
+        state_machine: Optional[StateMachine] = None,
+        add_output_event_task: bool = True,
+    ) -> None:
+        """
+        Create state machine. Automatically adds a task at the end that publishes an event to EventBridge.
+
+        Parameters
+        ----------
+        id : str
+            Identifier of the state machine
+        environment_id : str
+            Identifier of the environment
+        definition : IChainable
+            State machine definition
+        state_machine_type : StateMachineType
+            Type of the state machine
+        state_machine : Optional[StateMachine]
+            Escape hatch to override state machine
+        add_output_event_task : bool
+            Add task to publish an event at the end of the state machine. `True` by default.
+        """
+        if add_output_event_task:
+            definition = definition.next(self.get_event_task(self, f"{id}-event-task", [self.get_output_event()]))
```

Comment:
I am confused, if I provide an escape hatch, the definition I provide will not be used, yet we still try to add an output event to it?

Reply:
You'd have to pass `add_output_event_task=False` in this case but I'm ok to not do it if escape hatch is used

Reply:
It's more that if you are making `definition` optional when an escape hatch is provided and `add_output_event_task` is `True` by default, then that line would fail no?